File: theta_13tev_global/TxtDatabase.cpp
#ifndef txt_database_hh
#define txt_database_hh 1

/////////////////////////////////////////////////////////////////////////////
//  Author      :     P.Mandrik, IHEP
//  Date        :     02/09/17
//  Version     :     0.5.0
/////////////////////////////////////////////////////////////////////////////
//
//  Python module TxtDatabase.py provide class for write and read
//  access into simple human readeble txt database:
//
//    order : key : comment : list of data as single string with commas
//    order : key : comment : list of data as single string with commas
//    order : key : comment : list of data as single string with commas
//    
/////////////////////////////////////////////////////////////////////////////
//  Changelog : 
//    02/09/17 v0.5.0
//    days of creation, version 0.5.0
/////////////////////////////////////////////////////////////////////////////

static void ltrim(std::string &s) {
  s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](int ch) { return !std::isspace(ch); }));
}

static void rtrim(std::string &s) {
  s.erase(std::find_if(s.rbegin(), s.rend(), [](int ch) { return !std::isspace(ch); }).base(), s.end());
}

std::string strip(std::string s) {
  ltrim(s);
  rtrim(s);
  return s;
}

vector<string> * split_and_strip_string(string str, const string & sep = " "){
  int sep_size = sep.size();
  vector<string> * answer = new vector<string>();

  for(int i = 0; i + sep_size < str.size(); i++){
    if( sep == str.substr(i, sep_size) ){
      answer->push_back( strip( string(str.substr(0, i)) ) ); 
      str = str.substr(i + sep_size, str.size() - i - sep_size);
      i = 0;
    }
  }

  answer->push_back(str);
  return answer;
}

struct TxtData{
  TxtData(){}
  TxtData(std::string v, std::string c, int o) : value(v), comment(c), order(o) {}
  std::string value;
  std::string comment;
  std::string key;
  int order;
};

class TxtDatabase {
  public:
  TxtDatabase(){ next_order = 0; }
  TxtDatabase(std::string file_name){
    fname = file_name;
    next_order = 0;
    Read();
  }

  void Print(){
    std::vector<TxtData> data_list;
    for(auto item : data){  
      item.second.key = item.first;
      data_list.push_back( item.second );
    }
    sort(data_list.begin(), data_list.end(), [] ( const TxtData &a, const TxtData &b ) { return a.order < b.order; } );

    for(auto item : data_list) 
      cout << to_string(item.order) << " : " << item.key << " : " << item.comment << " : " << item.value << endl;
  }

  void Read(){
    std::string line;
    ifstream ifile ( fname.c_str() );
    if(ifile.is_open()){
      while( getline(ifile,line) ){
        // cout << line << '\n';
        auto items = split_and_strip_string(line, ":");
        if(items->size() != 4) { delete items; continue; }

        TxtData new_data;
        new_data.order   =  atoi( items->at(0).c_str() );
        new_data.key     =  strip(items->at(1)) ;
        new_data.comment =  strip(items->at(2)) ;
        new_data.value   =  strip(items->at(3)) ;

        data[ new_data.key ] = new_data;

        delete items;
        next_order = max(next_order, new_data.order);
      }
      ifile.close();
      return;
    }
  }

  vector<TxtData> GetMatch(string include_rexp, string exclude_rexp = ""){
    vector<TxtData> answer;

    regex reg_incl( include_rexp );
    regex reg_excl( exclude_rexp );

    for(auto item : data){
      if( !regex_match(item.first, reg_incl) ) continue;
      if(  regex_match(item.first, reg_excl) ) continue;

      answer.push_back( item.second );
    }
    return answer;
  }

  void Write(){
    ofstream ofile;
    ofile.open( fname.c_str() );
    std::vector<TxtData> data_list;
    for(auto item : data){  
      item.second.key = item.first;
      data_list.push_back( item.second );
    }
    sort(data_list.begin(), data_list.end(), [] ( const TxtData &a, const TxtData &b ) { return a.order < b.order; } );

    for(auto item : data_list) 
      ofile << to_string(item.order) << " : " << strip(item.key) << " : " << strip(item.comment) << " : " << strip(item.value) << endl;

    ofile.close();
  }

  std::string GetItem(const std::string & key){
    auto item = data.find( key );
    if(item != data.end()) return item->second.value;
    cout << "TxtDatabase.GetItem() " << fname << " wrong key = " << key << endl;
    return "";
  }

  std::string GetComment(const std::string & key){
    auto item = data.find( key );
    if(item != data.end()) return item->second.value;
    cout << "TxtDatabase.GetComment() " << fname << " wrong key = " << key << endl;
    return "";
  }

  void SetItem(const std::string & key, const std::string & value, const std::string & comment=""){
    auto item = data.find( key );
    if(item != data.end() ){
      int order = item->second.order;
      data[key] = TxtData(value, comment, order);
      return;
    }
    data[key] = TxtData(value, comment, next_order++);
  }
  
  map<std::string, TxtData> data;

  private:
  std::string fname;
  int next_order;
};

#endif

File: theta_13tev_global/TxtDatabase.py
# -*- coding: utf-8 -*-

"""
/////////////////////////////////////////////////////////////////////////////
  Author      :     P.Mandrik, IHEP
  Date        :     02/09/17
  Version     :     0.5.0
/////////////////////////////////////////////////////////////////////////////

  Python module TxtDatabase.py provide class for write and read
  access into simple human readeble txt database:

    order : key : comment : list of data as single string with commas
    order : key : comment : list of data as single string with commas
    order : key : comment : list of data as single string with commas
    
/////////////////////////////////////////////////////////////////////////////
  Changelog : 
    02/09/17 v0.5.0
    days of creation, version 0.5.0
"""

class TxtDatabase():
  def __init__(self, file_name):
    self.fname = file_name;
    self.data  = {}
    self.next_order = 0

  def Read(self):
    ifile         = open(self.fname, "r")
    raw_data = ifile.read()

    for line in raw_data.split("\n"):
      if not line or len( line ) < 3 : continue
      sublines = line.split(":")
      try:
        order = int(sublines[0])
        self.next_order = max(order, self.next_order)
        key = sublines[1].strip()
        comment = sublines[2].strip()
        value = sublines[3].strip()
        # print str(order)+"|"+key+"|"+comment+"|"+value
        self.data[key] = [value,comment, order]
      except : continue

  def Write(self):
    ofile = open(self.fname, "w")

    data_list = self.data.items()
    data_list.sort(key=lambda x: x[1][2], reverse=False) # sort by order
    for item in data_list:
      ofile.write( str(item[1][2]) + " : " + str(item[0]) + " : " + str(item[1][1]) + " : " );
      if type(item[1][0]) == type(list()):
        db = [str(val) for val in item[1][0] ]
        ofile.write( " ".join(db) + " \n" );
      else : ofile.write( str(item[1][0]) + " \n" );
    ofile.close()
  
  def GetItem(self, key):
    answer = None
    try:
      answer = self.data.get(key)[0]
      if answer : return answer
    except: pass
    if not answer : print "TxtDatabase.GetItem():", self.fname, "wrong key = ", "\"" + key + "\""
    return None

  def GetComment(self, key):
    answer = None
    try:
      answer = self.data.get(key)[1]
      if answer : return answer
    except: pass
    if not answer : print "TxtDatabase.GetComment():", self.fname, "wrong key = ", "\"" + key + "\""
    return None

  def SetItem(self, key, value, comment=""):
    item = self.data.get(key)
    if item:
      order = item[2]
      self.data[key] = [value, comment, order]
      return
    order = self.next_order
    self.next_order += 1
    self.data[key] = [value, comment, order]
      

def test():
  print "TxtDatabase() usage test ... "

  print "write ... "
  db = TxtDatabase("tmp.txt");
  db.SetItem("a", 12313, "test()")
  db.SetItem("b", "123 132 123 123 123 123", "test()")
  db.SetItem("abc", [123, 1334, 551, 515123], "test()")
  db.Write()
  print db.data

  print "read ... "
  db.Read()
  print db.data

  print "read and write ... "
  db.Read()
  db.Write()
  print db.data

if __name__ == "__main__": 
  import argparse
  parser = argparse.ArgumentParser(description='Produce some theta .cfg datacards ... ')
  parser.add_argument('--file',  dest='file',    type=str, default='db.txt', help='database name')
  parser.add_argument('--key',   dest='key',    type=str, default='',       help='key to get'   )
  args = parser.parse_args()
  if args.key:
    db = TxtDatabase(args.file) 
    db.Read()
    print db.GetItem(args.key)
  else : test()



